geoana/gravity.py

```diff
@@ -0,0 +1,261 @@
+
+        Define the point mass.
+
+        >>> location = np.r_[0., 0., 0.]
+        >>> mass = 1.0
+        >>> simulation = PointMass(
+        >>>     mass=mass, location=location
+        >>> )
+
+        Now we create a set of gridded locations and compute the gravitational field.
+
+        >>> xyz = ndgrid(np.linspace(-5, 5, 10), np.linspace(-5, 5, 10), np.array([0]))
+        >>> g = simulation.gravitational_field(xyz)
+
+        Finally, we plot the gravitational field lines.
+
+        >>> fig = plt.figure(figsize=(4, 4))
+        >>> ax = fig.add_axes([0.15, 0.15, 0.8, 0.8])
+        >>> plot2Ddata(xyz[:, 0::2], g[:, 0::2], ax=ax, vec=True, scale='log')
+        >>> ax.set_xlabel('x')
+        >>> ax.set_ylabel('y')
+        >>> ax.set_title('Gravitational field at z=0')
+        """
+
+        r_vec = xyz - self.location
+        r = np.linalg.norm(r_vec, axis=-1)
+        g_vec = (G * self.mass * r_vec) / r[..., None]
+        return g_vec
+
+    def gravitational_gradient(self, xyz):
+        """
+        Gravitational gradient for a point mass.
+
+        Parameters
+        ----------
+        xyz : (..., 3) numpy.ndarray
+            point mass location
+
+        Returns
+        -------
+        (..., 3, 3) numpy.ndarray
+            gravitational gradient at point mass location xyz
+
+        Examples
+        --------
+        Here, we define a point mass with mass=1kg and plot the gravitational
+        gradient.
+
+        >>> import numpy as np
+        >>> import matplotlib.pyplot as plt
+        >>> from geoana.gravity import PointMass
+        >>> from geoana.utils import ndgrid
+
+        Define the point mass.
+
+        >>> location = np.r_[0., 0., 0.]
+        >>> mass = 1.0
+        >>> simulation = PointMass(
+        >>>     mass=mass, location=location
+        >>> )
+
+        Now we create a set of gridded locations and compute the gravitational gradient.
+
+        >>> xyz = ndgrid(np.linspace(-1, 1, 20), np.linspace(-1, 1, 20), np.linspace(-1, 1, 20))
+        >>> g_tens = simulation.gravitational_gradient(xyz)
+
+        Finally, we plot the gravitational gradient of each element of the 3 x 3 matrix.
+
+        >>> fig, (ax1, ax2, ax3) = plt.subplots(1, 3, figsize=(12,6))
+        >>> ax1.set_aspect('equal')
+        >>> ax1.set_title('1')
+        >>> cf1 = ax1.contourf(xyz, xyz, g_tens[:,0,:], cmap='jet')
+        >>> fig.colorbar(cf1, ax=ax1)
+        >>> ax2.set_aspect('equal')
+        >>> ax2.set_title('2')
+        >>> cf2 = ax2.contourf(xyz, xyz, g_tens[:,1,:], cmap='jet')
+        >>> fig.colorbar(cf2, ax=ax2)
+        >>> ax3.set_aspect('equal')
+        >>> ax3.set_title('3')
+        >>> cf3 = ax3.contourf(xyz, xyz, g_tens[:,2,:], cmap='jet')
+        >>> fig.colorbar(cf3, ax=ax3)
+        >>> plt.show()
+        """
+
+        r_vec = xyz - self.location
+        r = np.linalg.norm(r_vec, axis=-1)
+        g_tens = (G * self.mass * np.eye(3)) / r[..., None, None] ** 3 +\
```

Comment:
I'm just now realizing I was a bit incomplete with what I wrote before for the formula for this one,

it should be:
`g_tens = G * mass * (eye / r**3 + 3 * outer(r_vec, r_vec)/r**5 ) `

the `G` and `mass` both need to multiple the entire bit containing all the geometric terms in there